cpmpy/solvers/pumpkin.py

```diff
@@ -0,0 +1,531 @@
+            for cpm_var in self.user_vars:
+                cpm_var._value = None
+
+        return has_sol
+
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+
+        if is_num(cpm_var):
+            return cpm_var
+
+        # special case, negative-bool-view
+        # work directly on var inside the view
+        if isinstance(cpm_var, NegBoolView):
+            return self.solver_var(cpm_var._bv).negate()
+
+        # create if it does not exist
+        if cpm_var not in self._varmap:
+            if isinstance(cpm_var, _BoolVarImpl):
+                revar = self.pum_solver.new_boolean_variable(name=str(cpm_var))
+            elif isinstance(cpm_var, _IntVarImpl):
+                revar = self.pum_solver.new_integer_variable(cpm_var.lb, cpm_var.ub, name=str(cpm_var))
+            else:
+                raise NotImplementedError("Not a known var {}".format(cpm_var))
+            self._varmap[cpm_var] = revar
+
+        # return from cache
+        return self._varmap[cpm_var]
+
+    def objective(self, expr, minimize=True):
+        """
+            Post the given expression to the solver as objective to minimize/maximize
+
+            - expr: Expression, the CPMpy expression that represents the objective function
+            - minimize: Bool, whether it is a minimization problem (True) or maximization problem (False)
+
+            'objective()' can be called multiple times, only the last one is stored
+
+            .. note::
+                technical side note: any constraints created during conversion of the objective
+                are premanently posted to the solver
+        """
+        # make objective function non-nested
+        obj_var = intvar(*get_bounds(expr))
+        self += expr == obj_var
+
+        # make objective function or variable and post
+        self._objective = obj_var
+        self.objective_is_min = minimize
+
+    def has_objective(self):
+        return self._objective is not None
+
+    # `__add__()` first calls `transform()`
+    def transform(self, cpm_expr):
+        """
+            Transform arbitrary CPMpy expressions to constraints the solver supports
+
+            Implemented through chaining multiple solver-independent **transformation functions** from
+            the `cpmpy/transformations/` directory.
+
+            See the 'Adding a new solver' docs on readthedocs for more information.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: list of Expression
+        """
+        # apply transformations
+        cpm_cons = toplevel_list(cpm_expr)
+        supported = {"alldifferent", "cumulative", 
+                     "min", "max", "element"}
+        cpm_cons = decompose_in_tree(cpm_cons, supported=supported) # TODO: Pumpkin supports half-reified globals
```

Comment:
I'm not sure I understand the TODO comment

Reply:
Currently we decompose any nested globals, in the future we should update the decompose_in_tree transformation to extract the ones in positive context to half-reifications.
Then, we can avoid a lot of the decompositions when we're doing e.g., MUS-extraction with assumptions